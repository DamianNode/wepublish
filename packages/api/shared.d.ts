export * from './lib/esm/shared'
