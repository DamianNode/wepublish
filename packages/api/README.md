# GraphQL Schema

```graphql
type Article {
  id: ID!
  title: String!
  lead: String!
  peer: Peer @relation(name: "Peer")
}

type Peer {
  id: ID!
  name: string
  url: string
}
```
