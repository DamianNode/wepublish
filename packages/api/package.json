{
  "name": "@wepublish/api",
  "version": "0.0.0",
  "main": "./lib/",
  "license": "MIT",
  "scripts": {
    "build": "run-s copy-license sync-package-info build:*",
    "build:typescript": "tsc -b ./src/tsconfig.json",
    "watch": "npm-run-all copy-license sync-package-info --parallel watch:*",
    "watch:typescript": "tsc -b ./src/tsconfig.json --watch",
    "copy-license": "shx cp ../../LICENSE ./",
    "clean": "shx rm -rf ./LICENSE ./lib",
    "prepare": "run-s clean copy:license build sync-package-info",
    "sync-package-info": "node ./../../scripts/syncPackageInfo.js ./src/version.ts"
  },
  "dependencies": {
    "graphql": "^14.4.2",
    "express-graphql": "^0.9.0"
  },
  "devDependencies": {
    "@types/node": "^12.6.8",
    "@types/graphql": "^14.2.3",
    "typescript": "^3.5.2",
    "npm-run-all": "^4.1.5",
    "shx": "^0.3.2"
  }
}
