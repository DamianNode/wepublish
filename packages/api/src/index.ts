import {RequestListener} from 'http'
import {GraphQLSchema, GraphQLObjectType, GraphQLString} from 'graphql'
import graphqlHTTP from 'express-graphql'

export interface Adapter {
  getArticle(): any
  getArticles(): any
  createArticle(): any
  updateArticle(): any
}

export interface HandlerOptions {
  adapter: Adapter
}

export function createAPIHandler(opts: HandlerOptions): RequestListener {
  const schema = new GraphQLSchema({
    query: new GraphQLObjectType({
      name: 'Root',
      fields: {
        test: {
          type: GraphQLString,
          resolve() {
            return 'test'
          }
        }
      }
    })
  })

  return graphqlHTTP({
    schema,
    graphiql: true
  })
}

export default createAPIHandler
