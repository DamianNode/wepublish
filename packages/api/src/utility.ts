import generateNanoID from 'nanoid/generate'

export const idAlphabet = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'

export function generateID() {
  return generateNanoID(idAlphabet, 10)
}
