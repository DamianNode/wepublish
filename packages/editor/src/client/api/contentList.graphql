query ContentList(
  $type: contentTypeEnum!
  $context: contentContextEnum
  $filter: String
  $after: ID
  $first: Int
) {
  content {
    _all {
      list(type: $type, context: $context, first: $first, after: $after, filter: {title: $filter}) {
        nodes {
          peer {
            id
            name
          }
          content {
            ...ContentListRef
          }
        }
        pageInfo {
          startCursor
          endCursor
          hasNextPage
          hasPreviousPage
        }
        totalCount
      }
    }
  }
}

query ContentGet($id: ID!) {
  content {
    _all {
      read(id: $id) {
        ...ContentListRef
      }
    }
  }
}

fragment ContentListRef on ContentModelSummary {
  id
  title
  shared
  contentType
  revision
  state
  createdAt
  modifiedAt
  publicationDate
  dePublicationDate
}

mutation PublishContent(
  $id: ID!
  $publishAt: DateTime!
  $publishedAt: DateTime!
  $updatedAt: DateTime!
) {
  content {
    _all {
      publish(id: $id, publishAt: $publishAt, publishedAt: $publishedAt, updatedAt: $updatedAt) {
        id
        pending {
          publishAt
        }
        published {
          publishedAt
        }
      }
    }
  }
}

mutation UnpublishContent($id: ID!) {
  content {
    _all {
      unpublish(id: $id) {
        id
        pending {
          publishAt
        }
        published {
          publishedAt
        }
      }
    }
  }
}

mutation DeleteContent($id: ID!) {
  content {
    _all {
      delete(id: $id)
    }
  }
}
