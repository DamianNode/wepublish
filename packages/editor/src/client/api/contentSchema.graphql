query Config {
  config {
    content {
      id
      identifier
      namePlural
      nameSingular
      schema
    }
    languages {
      defaultLanguageId
      languages {
        id
        tag
        description
      }
    }
  }
}
