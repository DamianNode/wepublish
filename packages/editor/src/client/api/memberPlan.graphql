fragment MemberPlanRef on MemberPlan {
    id
    label
    isActive
    image {
        ...ImageRef
    }
}

fragment FullMemberPlan on MemberPlan {
    description
    pricePerMonthMinimum
    pricePerMonthMaximum
    availablePaymentMethods {
        paymentMethod {
            ...FullPaymentMethod
        }
        paymentPeriodicity {
            id
            checked
        }
        minimumDurationMonths
        forceAutoRenewal
    }
    ...MemberPlanRef
}

query MemberPlanList($filter: String, $after: ID, $before: ID, $first: Int, $last: Int) {
    memberPlans(filter: {label: $filter}, after: $after, before: $before, first: $first, last: $last) {
        nodes {
            ...FullMemberPlan
        }

        pageInfo {
            startCursor
            endCursor
            hasNextPage
            hasPreviousPage
        }

        totalCount
    }
}

query MemberPlan($id: ID!) {
    memberPlan(id: $id) {
        ...FullMemberPlan
    }
}

mutation CreateMemberPlan($input: MemberPlanInput!) {
    createMemberPlan(input: $input) {
        ...FullMemberPlan
    }
}

mutation UpdateMemberPlan($id: ID!, $input: MemberPlanInput!) {
    updateMemberPlan(id: $id, input: $input) {
        ...FullMemberPlan
    }
}

mutation DeleteMemberPlan($id: ID!) {
    deleteMemberPlan(id: $id)
}
