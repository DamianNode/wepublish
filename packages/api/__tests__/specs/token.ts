import {MongoDBAdapter} from '@wepublish/api-db-mongodb'
import {ApolloServerTestClient} from 'apollo-server-testing'
import {createGraphQLTestClientWithMongoDB} from '../utility'
import {CreateToken, TokenList, DeleteToken} from '../api/private'

let testClientPublic: ApolloServerTestClient
let testClientPrivate: ApolloServerTestClient
let dbAdapter: MongoDBAdapter

beforeAll(async () => {
  try {
    const setupClient = await createGraphQLTestClientWithMongoDB()
    testClientPublic = setupClient.testClientPublic
    testClientPrivate = setupClient.testClientPrivate
    dbAdapter = setupClient.dbAdapter

    console.log('public', testClientPublic)
  } catch (error) {
    console.log('Error', error)
    throw new Error('Error during test setup')
  }
})

describe('Tokens', () => {
  describe('can be created/updated/edited/deleted:', () => {
    const ids: string[] = []
    beforeEach(async () => {
      const {mutate} = testClientPrivate
      const res = await mutate({
        mutation: CreateToken,
        variables: {
          input: {
            name: `Test Token Name ${ids.length}`
          }
        }
      })
      ids.unshift(res.data?.createToken?.id)
    })

    test('can be created', async () => {
      const {mutate} = testClientPrivate
      const res = await mutate({
        mutation: CreateToken,
        variables: {
          input: {
            name: 'New Create Token'
          }
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          createToken: {
            id: expect.any(String),
            token: expect.any(String)
          }
        }
      })
      ids.unshift(res.data?.createToken?.id)
      expect(res.data?.createToken.token).toBeDefined()
    })

    test('can be read in list', async () => {
      const {query} = testClientPrivate
      const res = await query({query: TokenList})
      expect(res).toMatchSnapshot({
        data: {
          tokens: Array.from({length: ids.length}, () => ({
            id: expect.any(String)
          }))
        }
      })
      expect(res.data?.tokens?.length).toBe(ids.length)
    })

    test('can be deleted', async () => {
      const {mutate} = testClientPrivate
      const res = await mutate({
        mutation: DeleteToken,
        variables: {
          id: ids[0]
        }
      })
      expect(res).toMatchSnapshot({
        data: {
          deleteToken: expect.any(String)
        }
      })
      expect(res.data?.deleteToken).toBe(ids[0])
      ids.shift()
    })
  })
})

afterAll(async () => {
  if (dbAdapter) {
    await dbAdapter.db.dropDatabase()
    await dbAdapter.client.close()
  }
})
