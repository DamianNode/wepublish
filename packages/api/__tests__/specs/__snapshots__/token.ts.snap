// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tokens can be created/updated/edited/deleted: can be created 1`] = `
Object {
  "data": Object {
    "createToken": Object {
      "id": Any<String>,
      "name": "New Create Token",
      "token": Any<String>,
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tokens can be created/updated/edited/deleted: can be deleted 1`] = `
Object {
  "data": Object {
    "deleteToken": Any<String>,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tokens can be created/updated/edited/deleted: can be read in list 1`] = `
Object {
  "data": Object {
    "tokens": Array [
      Object {
        "id": Any<String>,
        "name": "Test Token Name 2",
      },
      Object {
        "id": Any<String>,
        "name": "New Create Token",
      },
      Object {
        "id": Any<String>,
        "name": "Test Token Name 0",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
